src/v/datalake/partition_spec_parser.cc

```diff
@@ -10,260 +10,516 @@
+    parse_result<ss::sstring> parse_identifier(position pos) {
+        auto res = parse_simple_identifier(pos);
+        if (!res) {
+            res = parse_quoted_w_doubling<'`'>(pos);
+        }
+        return res;
     }
-    unparsed = source->unparsed;
 
-    skip_space(unparsed);
-    if (!skip_expected(unparsed, ")")) {
-        ctx.report_expected(unparsed, "')'");
-        return std::nullopt;
+    template<
+      typename Elem,
+      template<typename...>
+      class Container,
+      fixed_string Delim>
+    parse_result<Container<Elem>> parse_separated_list(
+      parse_result<Elem> (self::*elem_parser)(position), position pos) {
+        Container<Elem> result;
+        while (true) {
+            if (!result.empty()) {
+                auto delim_res = expect_with_ws<Delim>(pos);
+                if (!delim_res) {
+                    return {std::move(result), pos};
+                }
+                pos = delim_res.after();
+            }
+
+            auto elem_res = (this->*elem_parser)(pos);
+            if (!elem_res) {
+                return elem_res;
+            }
+            pos = elem_res.after();
+            result.push_back(std::move(elem_res).value());
+        }
     }
 
-    auto result = transform_field{
-      .source = std::move(source->val),
-      .transform = transform,
-    };
+    parse_result<iceberg::unresolved_partition_spec::column_reference>
+    parse_column_reference(position pos) {
+        return parse_separated_list<ss::sstring, std::vector, ".">(
+          &self::parse_identifier, pos);
+    }
 
-    return parse_result<transform_field>{
-      .val = std::move(result),
-      .unparsed = unparsed,
+    parse_result<iceberg::unresolved_partition_spec::field>
+    parse_column_reference_expr(position pos) {
+        auto res = parse_column_reference(pos);
+        if (!res) {
+            return res;
+        }
+        pos = res.after();
+        return {{.source_name = std::move(res).value()}, pos};
     };
-}
 
-std::optional<parse_result<iceberg::unresolved_partition_spec::field>>
-parse_partition_field(const std::string_view& str, parse_ctx& ctx) {
-    auto unparsed = str;
-    skip_space(unparsed);
-
-    transform_field tf;
-    if (auto parsed_tf = parse_transform_field(unparsed, ctx); parsed_tf) {
-        tf = std::move(parsed_tf->val);
-        unparsed = parsed_tf->unparsed;
-    } else if (auto parsed_sf = parse_qualified_identifier(unparsed);
-               parsed_sf) {
-        tf.source = std::move(parsed_sf->val);
-        tf.transform = iceberg::identity_transform{};
-        unparsed = parsed_sf->unparsed;
-    } else {
-        ctx.report_expected(unparsed, "qualified identifier or transform");
-        return std::nullopt;
+    using transform_ident = variant_of_identities<iceberg::transform>;
+
+    parse_result<transform_ident> parse_transform_name(position pos) {
+        auto tagged_res = parse_result<transform_ident>::dummy_failure();
```

Comment:
would this commit be easier to understand if you just added all new code and removed the old code? right now all the add/removes are jumbled together.

Reply:
My change was exactly adding new code and removing old code. However, to split this commit into two separate ones I'll need to introduce some tricks to ensure it is compilable in between them.

How the change is presented is a matter of the diff engine, not an inherent property of the change. Maybe could you just click "view file" in the per-file menu in the top-right corner to view the resulting code only?

That's not unlike separating code changes from formatting changes: does it make sense if Github has an whitespace-ignoring comparison mode?